### PR TITLE
Nullable annotate for Microsoft.Win32.Registry

### PR DESCRIPTION
**Pull request by manne**
Contributes to #2339

### CODE REVIEW

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (danmosemsft) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: I guess junk can be a discard (`_`) if you are doing cleanup.

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (danmosemsft) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: I defer to the reviewers who are more familiar with nullability annotations. But it seems a shame to have to sprinkle `!` on just about every use of `_hkey` when as far as I can see it's guaranteed to not be null unless a Dispose has happened. And I assume (but did not check) that every public codepath is protected by `EnsureNotDisposed();` either directly or indirectly. I guess the compiler can't analyze that deeply. @stephentoub is this the cleanest way to do it?

Although I feel like we already had this discussion about Dispose() in a previous PR.

Comment (stephentoub) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: If something is initialized in the ctor and only set to null on Dispose, we've generally made the field non-nullable.  If you're violating the contract by using the instance after Dispose, then upholding the contract isn't particularly important.

Comment (danmosemsft) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: Cool. I see it's also set to null in `Win32Error(..)` but it immediately throws. It's basically a variation of dispose but where the handle is invalid.

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (manne) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: Ok. There are some small things I can take care of:
A) To remove unneeded usings
B) To inline some variable declarations
C) To simplify some 'default' expressions

In total ~10 spot (is it the correct term in english?).

Which refactorings should I do?

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (manne) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: Awesome.

for those who are interested in the documentation 😉
https://github.com/dotnet/runtime/blob/e041241bd94adca59a6ec90173e1d90599deed87/docs/coding-guidelines/api-guidelines/nullability.md#L53-L54

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (danmosemsft) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: I think it's fine to do simple, safe localized tweaks to align with the common style in this repo and don't make the review harder. Sounds like those would.

**Comment by eiriktsarpalis** [comment]
Please hold off from merging, Microsoft.Win32.Registry has a dependency on `System.Security.Principal.Windows`, whose nullable annotation is still WIP.

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs]: This should be nullable

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs]: Not sure why banged, there is null check above ...

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs]: Seems [`valueName`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.win32.registry.getvalue?view=netframework-4.8#remarks) is nullable,  technically `defaultValue` can be null too

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/Registry.cs]: Same here `valueName` is nullable

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: `(byte[]?)null` suggested instead of `!`

File: src/libraries/Microsoft.Win32.Registry/ref/Microsoft.Win32.Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/ref/Microsoft.Win32.Registry.cs]: Cannot comment on line 40, all `RegistryKey.CreateSubKey(...)` overloads above return nullable https://github.com/dotnet/runtime/blob/d1f4a8f710b6e1a7f2b0b61af357e3932bcd9772/src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs#L128

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs]: not needed `!`

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.cs]: Cannot write on the line 539  `GetValueKind(string? name)` accepts nullable

File: src/libraries/Microsoft.Win32.Registry/src/System/Security/AccessControl/RegistrySecurity.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/System/Security/AccessControl/RegistrySecurity.cs]: Arguments nullability should match with the `ExceptionFromErrorCode` delegate declaration
`protected internal delegate Exception ExceptionFromErrorCode(int errorCode, string? name, SafeHandle? handle, object? context);`

Comment (buyaa-n) [reply src/libraries/Microsoft.Win32.Registry/src/System/Security/AccessControl/RegistrySecurity.cs]: and return type nullability

**Comment by eiriktsarpalis** [comment]
`System.Security.Principal.Windows` annotations have been merged to master, could you please rebase your branch on top of the latest master?

**Comment by manne** [comment]
@eiriktsarpalis Have I done the correct thing?

**Comment by buyaa-n** [comment]
@manne please update files for Linux / Mac versions too, if you are using VS change your project context to Linux/OSX respectively to see the files in VS

**Comment by eiriktsarpalis** [comment]
There seem to be nullability related build errors on windows too. They might be related to the newly annotated dependencies.

**Comment by eiriktsarpalis** [comment]
Hi @manne, and thank you for your contribution. 

What is the status of this PR? As part of our ongoing annotation work we are looking to get this project done asap. Will you able to complete it soon? If not, would you mind if we took this over, either by creating a separate PR or pushing commits to your branch?

**Comment by manne** [comment]
@eiriktsarpalis I am working on it right now

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (eiriktsarpalis) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: Looks like the return type is annotated as nullable on account of the code path on line 121. But it seems clear by the assertion above that this is not expected to happen unless there is a bug. I think non-nullability should be asserted.

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]:
```suggestion
            int ret = Interop.Advapi32.RegQueryValueEx(_hkey, name, null, ref type, (byte[]?)null, ref datasize);
```

File: src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs
Comment (buyaa-n) [reply src/libraries/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs]: Good point, if this not return null then all `public RegistryKey CreateSubKey(...)` overloads not return null

File: src/libraries/Microsoft.Win32.Registry/ref/Microsoft.Win32.Registry.cs
Comment (buyaa-n) [inline src/libraries/Microsoft.Win32.Registry/ref/Microsoft.Win32.Registry.cs]: As per @eiriktsarpalis's comment https://github.com/dotnet/runtime/pull/2351/files#r378221011 all this `CreateSubKey(...)` overloads not return null

**Review by buyaa-n — APPROVED** [review]
Except @eiriktsarpalis's comment and updates related to that change overall looks great, thank you @manne

**Review by eiriktsarpalis — APPROVED** [review]

**Comment by eiriktsarpalis** [comment]
LGTM. If you could just fix the merge conflicts we should be ready to merge your changes.

Thank you!

**Comment by manne** [comment]
Thx

**Pull request merged**
